SimPEG/regularization/coupling.py

```diff
@@ -0,0 +1,744 @@
+
+
+        ..math::
+
+            \phi_c(\mathbf{m_1},\mathbf{m_2})
+
+            = \lambda \sum_{i=1}^{M} \|\nabla \mathbf{m_1}_i \times \nabla \mathbf{m_2}_i \|^2
+
+            = \sum_{i=1}^{M} \|\nabla \mathbf{m_1}_i\|^2 \ast \|\nabla \mathbf{m_2}_i\|^2
+                - (\nabla \mathbf{m_1}_i \cdot \nabla \mathbf{m_2}_i )^2
+
+            = \|\phi_{cx}\|^2 + \|\phi_{cy}\|^2 + \|\phi_{cz}\|^2 (optional strategy, not used in this script)
+
+
+        '''
+        m1 = self.map1*model
+        m2 = self.map2*model
+        m1, m2 = self.models([m1,m2])
+
+        if self.regmesh.mesh.dim == 2:
+            Dx_m1, Dy_m1 = self.calculate_gradient(m1)
+            Dx_m2, Dy_m2 = self.calculate_gradient(m2)
+
+            temp1 = Dx_m1**2 + Dy_m1**2
+            temp2 = Dx_m2**2 + Dy_m2**2
+            term1 = temp1.dot(temp2)
+
+            temp1 = Dx_m1*Dx_m2 + Dy_m1*Dy_m2
+            term2 = (np.linalg.norm(temp1))**2
+            result = term1 - term2
+
+            if normalized:
+                norms1, norms2, norms = self.gradient_applitude_inv(m1, m2, fltr=False)
+                temp1 = (Dx_m1*norms1)**2 + (Dy_m1*norms1)**2
+                temp2 = (Dx_m2*norms2)**2 + (Dy_m2*norms2)**2
+                term1 = temp1.dot(temp2)
+
+                temp1 = Dx_m1*Dx_m2*norms + Dy_m1*Dy_m2*norms
+                term2 = (np.linalg.norm(temp1))**2
+                result = term1 - term2
+
+        elif self.regmesh.mesh.dim == 3:
+            Dx_m1, Dy_m1, Dz_m1 = self.calculate_gradient(m1)
+            Dx_m2, Dy_m2, Dz_m2 = self.calculate_gradient(m2)
+
+            temp1 = Dx_m1**2 + Dy_m1**2 + Dz_m1**2
+            temp2 = Dx_m2**2 + Dy_m2**2 + Dz_m2**2
+            term1 = temp1.dot(temp2)
+
+            temp1 = Dx_m1*Dx_m2 + Dy_m1*Dy_m2 + Dz_m1*Dz_m2
+            term2 = (np.linalg.norm(temp1))**2
+            result = term1 - term2
+
+            if normalized:
+                norms1, norms2, norms = self.gradient_applitude_inv(m1, m2, fltr=False)
+                temp1 = (Dx_m1*norms1)**2 + (Dy_m1*norms1)**2 + (Dz_m1*norms1)**2
+                temp2 = (Dx_m2*norms2)**2 + (Dy_m2*norms2)**2 + (Dz_m2*norms2)**2
+                term1 = temp1.dot(temp2)
+
+                temp1 = Dx_m1*Dx_m2*norms + Dy_m1*Dy_m2*norms + Dz_m1*Dz_m2*norms
+                term2 = (np.linalg.norm(temp1))**2
+                result = term1 - term2
+
+        return 0.5*result
+
+    def deriv(self, model):
+        '''
+        Computes the Jacobian of the cross-gradient.
+
+        :param list of numpy.ndarray ind_models: [model1, model2,...]
+
+        :rtype: numpy.ndarray
+        :return: result: gradient of the cross-gradient with respect to model1, model2
+
+        .. math::
+
+            See tutorial or documentation.
+
+        '''
+        m1 = self.map1*model
+        m2 = self.map2*model
+        m1, m2 = self.models([m1,m2])
+
+        if self.regmesh.mesh.dim == 2:
+            Dx = self.regmesh.aveFx2CC.dot(self.regmesh.cellDiffx)
```

Comment:
I don't know the meaning of "stash the cell derivative matrices". What should I do here

Reply:
It would appear that the Dx and Dy are only dependancy on the regimes, which would be know once the `super().__init__` finishes. Thus I would suggest creating these matrices at the end of the `__init__` function, and stashing them on the object, thus the `CrossGradient`s initialization function would look like:

```
def __init__(self, mesh, indActive, mapping, **kwargs):
    super().__init__(mesh, indActive, mapping, **kwargs)
    regmesh = self.regmesh
    self._Gx = regmesh.aveFx2CC @ regmesh.cellDiffx
    self._Gy = regmesh.aveFy2CC @ regmesh.cellDiffy
    if regmesh.mesh.dim == 3:
        self._Gz = regmesh.aveFz2CC @ regmesh.cellDiffz
```

There for you could just grab the gradients you needed every time from `self._Gx` etc.

Reply:
done. Thanks for clarifying it!